keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2038 @@
+can serve as a reference.
+
+In addition, kube-scheduler can be configured to contact a resource driver
+directly as a scheduler extender. This can avoid the need to communicate the
+list of potential and unsuitable nodes through the apiserver:
+
+```
+# to be decided: extend existing versions (like pkg/scheduler/apis/config/v1beta2)
+# or add a new alpha version?
+type Extender struct {
+...
+       // ManagedResourceDrivers is a list of resource driver names that are managed
+       // by this extender. A pod will be sent to the extender on the Filter, Prioritize
+       // and Bind (if the extender is the binder) phases iff the pod requests at least
+       // one ResourceClaim for which the resource driver name in the corresponding
+       // ResourceClass is listed here. In addition, the builtin dynamic resources
+       // plugin will skip setting SuitableNodes for claims managed by the extender
+       // if the extender has a FilterVerb.
+       ManagedResourceDrivers []string
+```
+
+The existing extender plugin must check this field to decide when to contact
+the extender.
+
+The following extension points are implemented in the new plugin:
+
+#### Pre-filter
+
+This checks whether a Pod uses any ResourceClaims. If there are ResourceClaims
+with immediate binding that are not allocated yet, then the Pod will be marked
+as unschedulable at the moment.
+
+#### Filter
+
+This checks whether the given node has access to those ResourceClaims which
+were already allocated.
+
+For unallocated ResourceClaims with delayed allocation, only those nodes are
+filtered out that are explicitly listed in UnsuitableNodes or that don't match
+the optional AvailableOnNodes node filter.
+
+There are several
+reasons why such a deny list is more suitable than an allow list:
+- Nodes for which no information is available must pass the filter phase to be
+  included in the list that will be passed to the post-filter and to get copied
+  into the PotentialNodes field there.
+- A node can already be chosen while there is no information yet and, if
+  allocation for that node actually works, the Pod can get scheduled sooner.
+- Some resource drivers might not have any unsuitable nodes, for example
+  because they modify containers and that works on all nodes at all
+  times. Forcing such drivers to set an allow list would cause unnecessary
+  work.
+
+In its state for the Pod the scheduler plugin must remember when it rejected a
+node because of UnsuitableNodes. That information will be used in Post-filter
+to deallocate resources.
+
+#### Post-filter
+
+This is called when no suitable node could be found. If the Pod depends on ResourceClaims with delayed
+allocation, then deallocating one or more of these ResourceClaims may make the
+Pod schedulable after allocating the resource elsewhere. Therefore each
+ResourceClaim with delayed allocation is checked whether all of the following
+conditions apply:
+- allocated
+- not currently in use
```

Comment:
why would the scheduler be responsible for deallocating a ResourceClaim? Shouldn't that be the driver?

Reply:
The scheduler needs to trigger it. The actual deallocation is then done by the driver. It has to be that way because only the scheduler knows when a pod got stuck due to allocated resources.

Reply:
but who knows when the allocated resource is not currently in use?

Reply:
That's tracked in `ResourceClaim.Status.UsedBy`: if that field is empty, the resource is unused.